### PR TITLE
Adapt @pvandkyen's build-cache action

### PR DESCRIPTION
**Pull request by kaitj**
This wf sets up an environment for a given version of Python using a given project dependency manager. At the moment, this is only setup to work with Poetry.

Users will have to provide the desired Python version(s) and a reference ID for caching purposes. Optional inputs include the dependency manager desired (again, only Poetry at the moment) and indicate with a boolean flag whether the project library should also be installed (default: false).

@pvandyken - do you want to take a look when you get a moment? ~I have yet to actually test this out, just adapted from your composite action (on the list of TODOs prior to any merge).~ Tested on a fork of scattr, will wait until this is merged in release before full integration.

~EDIT: Realised why this was created as a composite action (which can't be called from different repositories) when I went to test this,. Initial thought is still to include a reusable workflow that builds and caches the environment, as well as a generic composite action (referenced in the reusable workflow) that would need to be deployed in each downstream repository for the sole purpose of making use of the cache. While not an ideal solution, I think this will still simplify the downstream actions a lot, but open to suggestions.~

~_Marked as draft due to realisation of composite action vs reusable workflow_~

### CODE REVIEW

**Comment by kaitj** [comment]
Hmm, it looks like this is only possible for orgs under a Github Enterprise account unless I am missing something?

**Comment by kaitj** [comment]
@pvandyken - you were right, false alarm. Can actually call composite actions from different repositories. One thing to note would be repository settings and if the owner / maintainer of the repo changes it to only allow actions to be run from within org or from specified users / orgs.

See https://github.com/kaitj/scattr/actions/runs/6301955760/job/17108293122 for an example of using the pyenv (open to a better name given there is an actual `pyenv` out there) setup workflow.

**Comment by pvandyken** [comment]
Maybe just like `install-python-project` or something for the name?

**Comment by kaitj** [comment]
I renamed it to `installPyProject` which is hopefully a little less contentious :sweat_smile:

File: .github/actions/action-setup_task-installPyProject/action.yml
Comment (pvandyken) [inline .github/actions/action-setup_task-installPyProject/action.yml:64]: What's the purpose of `${{ steps.setup-poetry.outcome }} == 'success'`? If setup-poetry fails, shouldn't the whole workflow fail?

File: .github/actions/action-setup_task-installPyProject/action.yml
Comment (pvandyken) [inline .github/actions/action-setup_task-installPyProject/action.yml:10]: I'm actually pretty sure you can access the `github` context in a composite action. I'll try in my own snakebids version of this and let you know. But I think that's better than this vague `reference` input (sorry, I hadn't thought of this when I made my own push earlier

**Review by pvandyken — CHANGES_REQUESTED** [review]

File: .github/actions/action-setup_task-installPyProject/action.yml
Comment (pvandyken) [reply .github/actions/action-setup_task-installPyProject/action.yml:10]: Indeed it does work. Check out the latest commit there to see what I did

File: .github/actions/action-setup_task-installPyProject/action.yml
Comment (kaitj) [reply .github/actions/action-setup_task-installPyProject/action.yml:64]: This is more future proofing than anything. Right now the workflow only works for poetry so your assessment of it failing is correct, but if this workflow ends up supporting `hatch` in the future the `setup-poetry` step would be skipped. In this situation, `outcome != success`, but the workflow wouldn't fail. But maybe we end up deciding we actually want separate workflows for setting up different dependency managers.

Comment (pvandyken) [reply .github/actions/action-setup_task-installPyProject/action.yml:64]: Gotcha, no I think a simple workflow makes sense

**Comment by pvandyken** [comment]
I'll wait for this to finish then switch snakebids over to using this

File: .github/actions/action-setup_task-installPyProject/action.yml
Comment (kaitj) [reply .github/actions/action-setup_task-installPyProject/action.yml:10]: I think you are right. I just ran into a context issue with the secrets, but that might just be because it is secrets and not inputs.

Comment (pvandyken) [reply .github/actions/action-setup_task-installPyProject/action.yml:10]: Correct, secrets must be explicitly passed in (similar to inputs, but a different syntax)

Comment (kaitj) [reply .github/actions/action-setup_task-installPyProject/action.yml]: > Indeed it does work. [Check out the latest commit there to see what I did](https://github.com/akhanf/snakebids/pull/341/commits/c3f28d4fd3ddba4d0ce12444d4b2dea62b50d2b1)

Great, this change has been pushed in 8154b7d

> Correct, secrets must be explicitly passed in (similar to inputs, but a different syntax)

Yes, this gets a little confusing when chaining reusable workflows since a downstream workflow could inherit secrets from an upstream workflow.

**Review by pvandyken — APPROVED** [review]
Cool, if you've verified it works already, I have no more comments!

**Comment by kaitj** [comment]
Sounds good, I'll merge this in in the AM. May hold off on an immediate release (`v0.2.0`). Looking to update the versioning workflows a little bit so its more flexible, but if you wanted to use this immediately, can reference the `main` branch once merged in.

**Pull request merged**
